### ISSUE TITLE
Feature Proposal: Add Retry Policies

### ISSUE THREAD

**Issue by arkoc** [canton7/RestEase#89]
Here is what I propose:

```
public interface IApiClient
{
    [Retry(typeof(ExponentialRetry))]
    Task UpdateAsync(int num);

    [Retry(typeof(LinearRetry(retryCount: 3)))]
    Task AddAsync(int num);

    [Retry(typeof(NoRetry))] // <- can be skipped
    Task<int> GetAll();
}
```
What you mind about this ?

P.S. Thank you for amazing library.

**Comment by arkoc** [comment]
Good catch at syntax error :)

I think, there is no need to invent bicycle again. So guys at Microsoft have implemented retry policies in Azure Storage Client Library. We can adopt them here. 

[Here](https://gauravmantri.com/2012/12/30/storage-client-library-2-0-implementing-retry-policies/) is blog post about that.

1. Yeah. I implemented retry policies for my project, and it was fun to unit tests.
2. All retry policies are going to implement `IRetryPolicy` which whould have function like `ShouldRetry`.
3. From above blog-post:

- “Non-retryable” exceptions are all 400 ( >=400 and <500) class exceptions (Bad gateway, Not Found, etc.) as well as 501 and 505. 
- All other exceptions are “retryable”. This includes client side timeouts.

**Comment by canton7** [comment]
Looks good.

Do you think we need a retry policy per method, or could we put one on the RestClient? That would mean the user could just instantiate and configure and instance of the right policy, rather than having to provide a type in an attribute.

Another other question is whether the user should be supplying a retry *policy*, or an object which actually performs the retries. The former case would have methods to control whether a retry should be attempted for a given error, and what the delay time would be. The latter would take a `Func<Task>`, or the info to send to `HttpClient.SendAsync`, and would return the result / throw an error.

**Comment by arkoc** [comment]
Initially I was thinking that user should decide in its own on what functions to set retry policy, but now I am thinking that it will be overhead. So, in my opinion:

1. Retry policies should be configured in `RestClient` once and they will work for all calls.
2. I think sticking with Interface -> Implementation will be more *clean* way. So user can provide his own implementation of `IRetryPolicy`. Methods can be more *faster* way for users, but the point is, that in most cases users gonna just use `Linear` or `Exponential` policies. And if they decide to implement something special, more likely it will be complex thing and zipping all that logic in one method will be headache.

**Comment by arkoc** [comment]
One question more:
Should we retry if `AllowAnyStatusCode` is set ?

In my opinion, we shouldn't, because by setting `AllowAnyStatusCode` to true, user indicates that any response status is *valid* status for him.

**Comment by canton7** [comment]
> Retry policies should be configured in RestClient once and they will work for all calls. (I can't think a really good argument, why user should decide for what function to retry and what functions not.)

Agreed. Some of the other user-configurable serializers, etc, accept information on the method which was called, so if a user *really* cares they can implement their own retryer which looks at the method which was called. 

> I think sticking with Interface -> Implementation will be more clean way. So user can provide his own implementation of IRetryPolicy. Methods can be more faster way for users, but the point is, that in most cases users gonna just use Linear or Exponential policies. And if they decide to implement something special, more likely it will be complex thing and zipping all that logic in one method will be headache.

(Just as an aside, use an abstract base class rather than an interface -- that way you can to make changes in a backwards-compatible way).

I'm still a little unclear on whether you prefer the user specifying a retry *policy* (a class which describes when to retry and how long to wait before retrying again), or a class which actually does the retrying. In the former case we keep calling the user's class saying "OK, we got this response, do you want to retry?", in the latter case we call the user's class saying "OK here's the request. Let me know when you've finished retrying".

It's quite possible we could get the best of both worlds - the user-supplied class is just given a request and told "here, perform this request until it succeeds", but we provide a base implementation which has abstract methods to figure out when to retry and for how long, and actually does the retry looping itself.

Whichever we do, we should also make sure that we let the user decide whether to retry based on the content of the headers and the response, not just its status code.

> Should we retry if AllowAnyStatusCode is set ?

Tricky, yeah. `AllowAnyStatusCode` is a bit ham-fisted, in that it doesn't distinguish between different sorts of failure. The *intention* is to allow things like treating a 404 as a valid response, but it goes beyond that. I'm tempted to say that we should still retry by default (if nothing else, it's one less rule to have to remember when using retries), but that this needs to be configurable on the retrier (e.g. by telling the retrier whether the `AllowAnyStatusCode` attribute was declared on the method responsible for the current request).

**Comment by arkoc** [comment]
> I'm still a little unclear on whether you prefer the user specifying a retry policy (a class which describes when to retry and how long to wait before retrying again), or a class which actually does the retrying. In the former case we keep calling the user's class saying "OK, we got this response, do you want to retry?", in the latter case we call the user's class saying "OK here's the request. Let me know when you've finished retrying".

Ah, now I get the question. So, I prefer for users to specify retry **policy** rather than retry logic itself, because basically retry looping is same, sending request is same.... As far as I understand in retrying workflow we have only 2 configurations:
- When to retry ?
- How long to retry ?

This both questions are answerable by policy, if we had configuration like `How to retry?` then it would be reasonable to let integration point to retrier. Or we can have both policies and retrier. But in my opinion its just making users life harder. I think it should be maximally easy for user to setup retry, even custom one.

So, here what I imagine for now:
`bool ShouldRetry(RetryContext retryContext, out TimeSpan retryInterval);`

`RetryContext` should contain, `HttpResponseMessage`, Method for which is request called and so on.

**Comment by canton7** [comment]
One other thing to bear in mind is that requests can time out, in which case HttpClient will throw a TaskCanceledException. We need to distinguish this from a user-cancelled request. This also means that we can't just give a `HttpResponseMessage to the user and say "does this need retrying?" - we also need to indicate that a request timed out.

That got me thinking, and I think there are actually two concepts here: whether a request was a success / failure / needs retrying, and the retry algorithm (linear, exponential, etc). The former is currently handled by a combination of `HttpResponseMessage.EnsureSuccessStatusCode` and `AllowAnyStatusCode`, but that's a bit coarse-grained (it doesn't distinguish between different classes of status code, and doesn't allow for "failures" caused by the body of a 200 response).

Maybe what we want is an object which decides whether a response is a success, a failure, or needs retrying.  If this indicates that a response was a success, we return it to the user. If it indicates that it was a failure, we throw an ApiException. If it indicates that it needs retrying, we pass it to the retry policy, which then decides whether to retry it or fail it (if there's no retry policy set, we automatically fail it). We'll also have to ensure that a timeout can never be considered a success, since there's no sensible response we can return to the client. This object is told whether `[AllowAnyStatusCode]` was set, and uses this to inform its decision.

The retry policy, then, just implements the retry algorithm. It gets told how many retries have occurred so far, and returns whether to retry again, and how long to wait. Again timeouts add complexity: if we have a linear timeout of 1 minute, and a request times out after 30 seconds, do we wait 30 seconds or 1 minute?

So to summarise (I haven't thought up any good names for these concepts and objects):

```csharp
enum RequestResult { Success, Fail, Retry }
abstract class FailurePolicy // Or something, I can't think of a good name
{
    // Gets information about the request which timed out (including whether AllowAnyStatusCode was set), decides whether to retry on timeout
    abstract bool RetryOnTimeout(SomeContext context);

    // Gets information about a request and a response
    abstract RequestResult ResultForResponse(SomeOtherContext context);
}

abstract class RetryPolicy
{
    // timeSinceLastRequest ties into the question on how long to wait on a timeout
    abstract TimeSpan? ShouldRetry(int retryCount, TimeSpan timeSinceLastRequest);
}
```

**Comment by arkoc** [comment]
I am a little unclear, why we need to introduce new entity like `RequestResult`. So If user want to retry if there is a custom header aet, he should define 2 separate classes, one for `RetryPolicy` and one for `FailurePolicy`. In my opinion it is getting harder than it should. 

For user who want to determine retry the call or not, we give him all contextual information, like ResponseMessage, thrown exception (if any)*, already retried count and so on. After that we expect from him to decide should we retry again or no.  (_Notice_: We should call `ShouldRetry` for any response from server, even if there was exception. )

* This also handles situation with TaskCancelledException.

**Comment by arkoc** [comment]
I agreed. Your approach will give more flexibility and in same time makes custom implementations more easier.

**Comment by canton7** [comment]
So really, RetryPolicy becomes more of a RetryAlgorithm. It really only needs to have a method which returns an `IEnumerable<TimeSpan>`.

**Comment by drmcclelland** [comment]
Polly (https://github.com/App-vNext/Polly) is a really great library with patterns like Retry & Circuit Breaker that would be a good combination to use with RestEase.

**Comment by canton7** [comment]
I've spent a bit of time reading the Polly docs.

Polly wants to work by taking over responsibility for executing your request. So the easiest way of "adding" retry support is to tell users "configure a Polly policy, then use it to execute your interface methods". However, that has the following issues:

1. Users have to wrap every call to an interface method. Fine, but a bit ugly.
2. People reading the Polly docs will expect to handle a HttpResponseMessage, HttpStatusCode, HttpRequestException, etc, but we will give them an ApiException or a strongly-typed result.

So perhaps a better system is to introduce a layer between the HttpResponseMessage and the deserialization which we let users hook into, which works in the world of HttpResponseMessages and lets users hook into their policy's Execute method. That means we'll have one policy for the whole API, which is good because the user doesn't need to repeat the policy for every request, but had because the user can't have per-request policies (which are needed for things like fallbacks).

We could let users pass a policy into an interface method, but I don't want to explicitly support Polly's types. There's also debatable difference between using a policy to execute an interface method, and passing a policy into an interface method to be executed.

Talking about fallbacks, we would want them to work with the deserialized object types, so we want to be working in concrete types and not HttpResponseMessages at this point. 

At the end of the day, RestEase has always been about making common operations easier, while complex operations stay about the same (returning a HttpResponseMessage, doing the deserialization manually, etc). So we need to figure out what the common Polly use-cases are, then focus on making those easier.

**Comment by canton7** [comment]
Also need to think about allowing Policy to cancel the request - should be easy.

Also PolicyResult -- I don't think there's a neat way we can handle this? Unless we somehow allow a `Task<PolicyResult>`?

Note that fallbacks only make sense for per-method policies - every method on the interface isn't going to return the same thing. We could potentially take a fallback, or delegate which returns a fallback, as a method parameter, and pass it though into the policy in some semi-sensible way? However we'd have to run a runtime check that the type of fallback was compatible with the type of result...

We also can't forget about `[AllowAnyStatusCode]`

**Comment by shatl** [comment]
+1 for using Polly with HttpClientFactory.

**Comment by clydevassallo** [comment]
Is it currently possible to register RestEase clients in the ServiceCollection using the [HttpClientFactory](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.ihttpclientfactory?view=dotnet-plat-ext-3.1) and [IHttpClientBuilder](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.ihttpclientbuilder?view=dotnet-plat-ext-3.1)?

**Comment by canton7** [comment]
It's not something I've played with, but I've heard of someone doing that quite easily. I've got plans to add a package to explicitly support that at some point: I imagine it will look a bit like [Refit's version](https://github.com/reactiveui/refit/blob/master/Refit.HttpClientFactory/HttpClientFactoryExtensions.cs)

**Comment by canton7** [comment]
Since you have a direct need right now, would you be able to put together an extension method which works for you, which I can use as a starting point? It's been a while since I used asp.net.

**Comment by clydevassallo** [comment]
I could definitely try. Let me give it a shot and open a PR.

**Comment by canton7** [comment]
Or even just post it in here. I'm working on a large-scale refactoring to support source generators at the moment, and I won't be able to take any other changes until that's complete.

**Comment by clydevassallo** [comment]
Will do. Whatever works best for you :)

**Comment by clydevassallo** [comment]
```cs
public static class ServiceCollectionExtensions
{
    private static readonly MethodInfo addRestEaseClientGenericMethodInfo = typeof(ServiceCollectionExtensions).GetTypeInfo().GetMethods().First(x => x.Name == "AddRestEaseClient" && x.IsStatic && x.GetParameters().Length == 1 && x.IsGenericMethod);

    public static IHttpClientBuilder AddRestEaseClient<T>(this IServiceCollection services) where T : class
    {
        return services
            .AddHttpClient(typeof(T).ToString())
            .AddTypedClient((client, serviceProvider) => RestClient.For<T>(client));
    }

    public static IHttpClientBuilder AddRestEaseClient(this IServiceCollection services, Type type)
    {
        if (type == null)
            throw new ArgumentNullException(nameof(type));

        var method = addRestEaseClientGenericMethodInfo.MakeGenericMethod(type);
        return (IHttpClientBuilder) method.Invoke(null, new object[]{ services });
    }
}
```

I created a simple initial version and it is working. I wrote some unit tests for basic cases and so far so good.

As you can see, this implementation hides access to the configuration of your RestClient. The next step would be to expose this configuration by adding a parameter to these extension methods with a type called RestEaseSettings. 

This new type would contain the following properties: 
1. JsonSerializerSettings
2. IResponseDeserializer
3. IRequestBodySerializer
4. RequestPathParamSerializer
5. IRequestQueryParamSerializer
6. QueryStringBuilder
7. IFormatProvider
If set by the caller, these properties will override the defaults in the RestClient.

I could continue developing this if you're interested in adding the functionality.

**Comment by canton7** [comment]
Thanks! I'll look at this when I've got some bandwidth.

**Comment by canton7** [comment]
@clydevassallo Do you think this looks sensible? https://github.com/canton7/RestEase/compare/develop...feature/httpclientfactory

**Comment by clydevassallo** [comment]
@canton7 Looks good :) I like the idea of having the configuration through an Action.

**Comment by canton7** [comment]
Closing with the next release: Polly is the preferred approach, and I've added docs to the README explaining how to use it with RestEase.
